### PR TITLE
config: Add api requests max & deadline configs

### PR DESCRIPTION
**Pull request by vadmeste**
## Description
Add two new configuration entries, api.requests-max and
api.requests-deadline which have the same role of
MINIO_API_REQUESTS_MAX and MINIO_API_REQUESTS_DEADLINE.

## Motivation and Context
Add new config to tune max api requests & deadline.

## How to test this PR?
Set max requests to 1 and issue parallel requests with minio-go

```go
package main

import (
	"bytes"
	"fmt"
	"log"
	"sync"

	"github.com/minio/minio-go/v6"
)

func main() {
	s3Client, err := minio.New("localhost:9000", "minio", "minio123", false)
	if err != nil {
		log.Fatalln(err)
	}

	wg := sync.WaitGroup{}
	ch := make(chan int)

	for i := 0; i < 20; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()

			by := []byte("AWPQ")
			for nr := range ch {
				_, err := s3Client.PutObject(
					"testbucket",
					fmt.Sprintf("object-%d", nr),
					bytes.NewReader(by),
					int64(len(by)),
					minio.PutObjectOptions{},
				)
				if err != nil {
					log.Fatalln(err)
				}
			}
		}()
	}

	for i := 0; i < 1000; i++ {
		ch <- i
	}

	close(ch)
	wg.Wait()
}
```

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [x] Documentation needed
- [ ] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#f8fd2d61-f1a6-4d6e-b17a-b50cfa9bcbf0) |
<a name="f8fd2d61-f1a6-4d6e-b17a-b50cfa9bcbf0"></a>
### 9273-d5c73c6/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:30452
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 88dcfed747ef:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 9 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 9 minutes and 27 seconds
(4/15) Running aws-sdk-ruby tests ... done in 19 seconds
(5/15) Running awscli tests ... done in 3 minutes and 6 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... done in 7 minutes and 2 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 54 seconds
(9/15) Running minio-go tests ... done in 20 minutes and 39 seconds
(10/15) Running minio-java tests ... done in 31 minutes and 25 seconds
(11/15) Running minio-js tests ... FAILED in 3 minutes and 59 seconds
{
  "name": "minio-js",
  "function": "listObjectsV2WithMetadata(bucketName, prefix, recursive, startAfter)",
  "args": "bucketName:minio-js-test-0d7a1509-719f-4512-b57c-49b55a6cd59a, recursive:true",
  "duration": 207,
  "status": "FAIL",
  "error": "TypeError: Cannot read property '0' of undefined at /mint/run/core/minio-js/node_modules/minio/dist/main/xml-parsers.js:414:42 at Array.forEach (<anonymous>) at parseListObjectsV2WithMetadata (node_modules/minio/dist/main/xml-parsers.js:409:21) at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:88:19) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}

Executed 10 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by vadmeste** [comment]
Not WIP anymore but we still need approval of naming & hierarchy from @abperiasamy

**Comment by vadmeste** [comment]
@harshavardhana I implemented your suggestion, registering S3 API is in the original place now.

File: docs/config/README.md
Comment (harshavardhana) [inline docs/config/README.md]: This should be `requests_max` we do not follow `-` anymore and the other one is `request_deadline`

File: docs/config/README.md
Comment (harshavardhana) [inline docs/config/README.md]:
```suggestion
api   manage global HTTP API call specific features, such as throttling, authentication types, etc.
```

File: docs/config/README.md
Comment (harshavardhana) [inline docs/config/README.md:190]: We should also edit `docs/throttle/` as well

File: cmd/config/api/api.go
Comment (harshavardhana) [inline cmd/config/api/api.go]:
```suggestion
	apiRequestsMax      = "requests_max"
```

File: cmd/config/api/api.go
Comment (harshavardhana) [inline cmd/config/api/api.go]:
```suggestion
	apiRequestsDeadline = "requests_deadline"
```

File: cmd/config-current.go
Comment (harshavardhana) [inline cmd/config-current.go]:
```suggestion
			Description: "manage global HTTP API call specific features, such as throttling, authentication types, etc.",
```

File: cmd/globals.go
Comment (harshavardhana) [inline cmd/globals.go:158]: Add a comment here

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_exclamation_mark: [more...](#8b9f20d7-1f65-4370-a10b-fb11a69f55a7) |
| mint-gateway-azure.sh | :x: [more...](#5d929380-f4a0-4364-b594-b32d6881eb22) |
<a name="5d929380-f4a0-4364-b594-b32d6881eb22"></a>
### 9273-fff7a58/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:30747
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b2adcfde5ace:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 9 minutes and 49 seconds
(4/15) Running aws-sdk-ruby tests ... done in 17 seconds
(5/15) Running awscli tests ... done in 3 minutes and 17 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 6 minutes and 38 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 32 seconds
(9/15) Running minio-go tests ... done in 19 minutes and 20 seconds
(10/15) Running minio-java tests ... done in 26 minutes and 47 seconds
(11/15) Running minio-js tests ... FAILED in 3 minutes and 46 seconds
{
  "name": "minio-js",
  "function": "listObjectsV2WithMetadata(bucketName, prefix, recursive, startAfter)",
  "args": "bucketName:minio-js-test-fd6b5e46-7b2d-4f63-b866-fd3a64c46d5d, recursive:true",
  "duration": 210,
  "status": "FAIL",
  "error": "TypeError: Cannot read property '0' of undefined at /mint/run/core/minio-js/node_modules/minio/dist/main/xml-parsers.js:414:42 at Array.forEach (<anonymous>) at parseListObjectsV2WithMetadata (node_modules/minio/dist/main/xml-parsers.js:409:21) at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:88:19) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}

Executed 10 out of 15 tests successfully.
```
<a name="8b9f20d7-1f65-4370-a10b-fb11a69f55a7"></a>
### 9273-fff7a58/mint-fs.sh.xtrace:
```
DISABLED TIME BEING
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#e5966413-1a01-4233-8eb4-3862099f563a) |
<a name="e5966413-1a01-4233-8eb4-3862099f563a"></a>
### 9273-82af03a/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:31102
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 0adcba5af46a:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 9 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 9 minutes and 46 seconds
(4/15) Running aws-sdk-ruby tests ... done in 18 seconds
(5/15) Running awscli tests ... FAILED in 1 minutes and 2 seconds
{
  "name": "awscli",
  "duration": 2077,
  "function": "aws --endpoint-url http://minio-c3.minio.io:31102 s3api create-multipart-upload --bucket \nAn error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it. --key \nAn error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it.-object\n",
  "status": "FAIL",
  "error": ""
}

Executed 4 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#2343a6a2-6145-4ff8-a425-39ad81b6df04) |
<a name="2343a6a2-6145-4ff8-a425-39ad81b6df04"></a>
### 9273-0359887/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c4.minio.io:31272
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp f39ab90cefad:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 9 minutes and 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 17 seconds
(5/15) Running awscli tests ... FAILED in 44 seconds
{
  "name": "awscli",
  "duration": 1482,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it."
}

Executed 4 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by vadmeste** [comment]
@harshavardhana all comments addressed, please take a look in the throttling documentation change

File: docs/throttle/README.md
Comment (harshavardhana) [inline docs/throttle/README.md]: There is no such thing as api.request_max @vadmeste either you need to use `mc admin config set` command fully and appropriately

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Apart from the doc typo, everything else looks good to me

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#67a1c688-3bb9-4257-b4c1-f5b946a45118) |
<a name="67a1c688-3bb9-4257-b4c1-f5b946a45118"></a>
### 9273-c672440/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:31313
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp ba42b9f9c332:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 1 minutes and 17 seconds
(4/15) Running aws-sdk-ruby tests ... done in 14 seconds
(5/15) Running awscli tests ... done in 2 minutes and 57 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... done in 6 minutes and 3 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 28 seconds
(9/15) Running minio-go tests ... done in 18 minutes and 48 seconds
(10/15) Running minio-java tests ... done in 26 minutes and 20 seconds
(11/15) Running minio-js tests ... FAILED in 4 minutes and 17 seconds
{
  "name": "minio-js",
  "function": "listObjectsV2WithMetadata(bucketName, prefix, recursive, startAfter)",
  "args": "bucketName:minio-js-test-f3a502f5-1274-4240-a05b-8d006f631357, recursive:true",
  "duration": 170,
  "status": "FAIL",
  "error": "TypeError: Cannot read property '0' of undefined at /mint/run/core/minio-js/node_modules/minio/dist/main/xml-parsers.js:414:42 at Array.forEach (<anonymous>) at parseListObjectsV2WithMetadata (node_modules/minio/dist/main/xml-parsers.js:409:21) at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:88:19) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}

Executed 10 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#40f6c1d2-1418-4c89-8549-2fb2d423e182) |
<a name="40f6c1d2-1418-4c89-8549-2fb2d423e182"></a>
### 9273-039c3aa/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:32055
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp ad2fc1fd4413:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 1 minutes and 22 seconds
(4/15) Running aws-sdk-ruby tests ... done in 14 seconds
(5/15) Running awscli tests ... done in 3 minutes and 1 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... done in 6 minutes and 25 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 30 seconds
(9/15) Running minio-go tests ... FAILED in 13 minutes and 10 seconds
{
  "args": {
    "destination": {},
    "sourceList": "source array of 10,001 elements"
  },
  "duration": 1651,
  "error": "The specified bucket is not valid.",
  "function": "ComposeObject(destination, sourceList)",
  "message": "PutObject failed",
  "name": "minio-go: testComposeObjectErrorCasesWrapper",
  "status": "FAIL"
}

Executed 8 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

**Review by kannappanr — APPROVED** [review]
LGTM

**Comment by kannappanr** [comment]
Need to check with @abperiasamy regarding naming and hierarchy before merging

**Review by Praveenrajmani — APPROVED** [review]
LGTM

**Comment by harshavardhana** [comment]
> Need to check with @abperiasamy regarding naming and hierarchy before merging

@kannappanr the env value was defined in this style already, we agreed that this is how it will be.

**Pull request merged**
